File: 0290_word_pattern.py
# n = len(pattern)
# Time: O(n)
# Time: O(n)

class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()

        if len(pattern) != len(words):
            return False
        if len(set(pattern)) != len(set(words)):
            return False

        pattern_to_word = {}

        for i in range(len(pattern)):
            if pattern[i] not in pattern_to_word:
                pattern_to_word[pattern[i]] = words[i]
            elif pattern_to_word[pattern[i]] != words[i]:
                return False

        return True

File: go/0290_word_pattern.go
package main

import "strings"

// n = len(pattern)
// m = len(s)
// Time: O(n+m)
// Space: O(n+m)
func wordPattern(pattern string, s string) bool {
	words := strings.Split(s, " ")

	if len(pattern) != len(words) {
		return false
	}

	p2w := make(map[byte]string)
	w2p := make(map[string]byte)

	for i := 0; i < len(words); i++ {
		p := pattern[i]
		w := words[i]

		if mappedWord, ok := p2w[p]; ok {
			if mappedWord != w {
				return false
			}
		} else {
			p2w[p] = w
		}

		if mappedPattern, ok := w2p[w]; ok {
			if mappedPattern != p {
				return false
			}
		} else {
			w2p[w] = p
		}
	}

	return true
}
